cmake/modules/FindZephyr-sdk.cmake

```diff
@@ -60,14 +60,41 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
       set(ZEPHYR_TOOLCHAIN_VARIANT ${ZEPHYR_CURRENT_TOOLCHAIN_VARIANT})
     endif()
   else()
-    find_package(Zephyr-sdk ${Zephyr-sdk_FIND_VERSION} REQUIRED QUIET CONFIG PATHS
-                 /usr
-                 /usr/local
-                 /opt
-                 $ENV{HOME}
-                 $ENV{HOME}/.local
-                 $ENV{HOME}/.local/opt
-                 $ENV{HOME}/bin)
+    # Paths that are used to find installed Zephyr SDK versions
+    SET(ZEPHYR_SDK_SEARCH_PATHS
+        /usr
+        /usr/local
+        /opt
+        $ENV{HOME}
+        $ENV{HOME}/.local
+        $ENV{HOME}/.local/opt
+        $ENV{HOME}/bin)
+
+    # Search for Zephyr SDK version 0.0.0 which does not exist, this is needed to
+    # return a list of compatible versions and find the best suited version that
+    # is available.
+    find_package(Zephyr-sdk 0.0.0 EXACT QUIET CONFIG PATHS ${ZEPHYR_SDK_SEARCH_PATHS})
+
+    # Remove duplicate entries and sort naturally in descending order.
+    list(REMOVE_DUPLICATES Zephyr-sdk_CONSIDERED_VERSIONS)
+    list(SORT Zephyr-sdk_CONSIDERED_VERSIONS COMPARE NATURAL ORDER DESCENDING)
+
+    # Loop over each found Zepher SDK version until one is found that is compatible.
+    foreach(ZEPHYR_SDK_CANDIDATE ${Zephyr-sdk_CONSIDERED_VERSIONS})
+      if("${ZEPHYR_SDK_CANDIDATE}" VERSION_GREATER_EQUAL "${Zephyr-sdk_FIND_VERSION}")
+        find_package(Zephyr-sdk ${ZEPHYR_SDK_CANDIDATE} EXACT QUIET CONFIG PATHS ${ZEPHYR_SDK_SEARCH_PATHS})
```

Comment:
this will cause you to loop from newest to oldest, like `3.0;2.0;1.0` so on first loop you invoke:
```
find_package(Zephyr-sdk 3.0 EXACT ...)
```
but of course the 3.0 installation of Zephyr SDK is compatible with requested version 3.0, so that shall always return true.

What you should do is to test each candidate against the `Zephyr-sdk_FIND_VERSION`, and in order to do so, you must use `Zephyr-sdk_CONSIDERED_CONFIGS` list, like this:
```
  foreach(zephyr_sdk_config ${Zephyr-sdk_CONSIDERED_CONFIGS})
    <adjust path of zephyr_sdk_config, to let's say zephyr_sdk_config_path>
    find_package(Zephyr-sdk ${Zephyr-sdk_FIND_VERSION} EXACT QUIET CONFIG PATHS <zephyr_sdk_config_path> NO_DEFAULT_PATH)
```
that way you'll test each candidate individually to see if it is compatible with the build system requested `Zephyr-sdk_FIND_VERSION`.

For inspiration on the `CONSIDERED_CONFIGS` path handling, find inspiration here:
https://github.com/zephyrproject-rtos/zephyr/blob/ca3a4f5f5b357c843d7f07bef443c6ca1945c249/share/zephyr-package/cmake/zephyr_package_search.cmake#L82-L85
Note: that code cannot be used as-is, but serve as inspiration for: `<adjust path of zephyr_sdk_config, ...>`

Reply:
Fixed